### PR TITLE
Add exact ratio distribution for deterministic test data generation

### PR DESCRIPTION
**Pull request by layoutd**
**NOTE:** This PR branches from #182 and will need merge conflict resolution with #184 and #185 (all modify the `batch()` method). The conflicts are trivial - just combining initialization calls.

Builds on: #182 

## Summary

Adds exact ratio distribution for coupon and refund generation in batch mode, eliminating variance from probabilistic distribution. Also includes critical bug fixes for weighted selection and counter restoration.

### Current Behavior (Probabilistic)
With 100 orders at 0.5 coupon ratio:
- ❌ Get 45-55 coupons (varies due to random chance)
- ❌ Can't predict exact counts for testing

### New Behavior (Deterministic)  
With 100 orders at 0.5 coupon ratio:
- ✅ Get exactly 50 coupons (pre-calculated distribution)
- ✅ Predictable counts for testing and verification

## Key Changes

1. **Exact Ratio Distribution** - Pre-generates boolean/integer arrays with exact counts before shuffling
2. **Fixed Weighted Selection Bug** - Corrected threshold calculation that was causing incorrect distribution
3. **Counter Restoration** - Restores `orders_remaining` and distribution arrays when order generation fails

## Benefits

1. **Deterministic Testing** - Exact ratios make test verification trivial
2. **No Variance** - Eliminates probabilistic randomness in batch operations
3. **Backwards Compatible** - Single orders still use probabilistic approach
4. **Memory Efficient** - Integer constants instead of strings for large batches
5. **Correct Distribution** - Fixed threshold bug ensures ratios work as expected

## Testing

```bash
# Test exact coupon ratio
ddev wp wc generate orders 100 --coupon-ratio=0.5 --status=completed
# Verify: Query wp_wc_orders_meta for _coupon_lines_hash, should be exactly 50

# Test exact refund ratio  
ddev wp wc generate orders 100 --status=completed --refund-ratio=0.4
# Verify: Should have exactly 20 full, 10 partial, 10 multi-partial refunds (40 total)

# Test counter restoration on failure
ddev wp wc generate orders 100 --coupon-ratio=0.5 --refund-ratio=0.4
# Intentionally cause some failures, verify counters restore correctly
```

<details>
<summary>Original Implementation Details</summary>

### Coupon Distribution
Pre-generates an array of boolean flags:
```php
// 100 orders, 0.5 ratio → exactly 50 true, 50 false
[true, true, ..., false, false]
shuffle($flags);
```

**Odd number handling:** Uses `round()` for non-exact splits:
- 11 orders @ 0.5 ratio → round(11 × 0.5) = round(5.5) = **6 with coupons, 5 without**

### Refund Distribution  
Pre-generates refund type flags with exact split using integer constants for memory efficiency:
```php
// Integer constants for memory optimization
const REFUND_TYPE_NONE = 0;
const REFUND_TYPE_FULL = 1;
const REFUND_TYPE_PARTIAL = 2;
const REFUND_TYPE_MULTI = 3;

// 100 orders, 0.4 refund ratio → 40 refunds total:
// - 20 full (50%)
// - 10 single partial (25%)
// - 10 multi-partial (25%)
[1, 1, ..., 2, 2, ..., 3, 3, ..., 0, 0, ...]
shuffle($flags);
```

**Odd number handling:** Remainder goes to multi-partial:
- 11 orders @ 0.4 ratio → 4 total refunds (rounded)
  - round(4 × 0.5) = **2 full**
  - round(4 × 0.25) = **1 partial**
  - 4 - 2 - 1 = **1 multi** (remainder)
  - 11 - 4 = **7 none**

**Memory efficiency:** For 20,000 orders, integers use ~160 KB vs strings ~160-320 KB, plus better CPU cache locality.

</details>

### CODE REVIEW

File: includes/Generator/Order.php [hunk 1]
```diff
@@ -790,4 +848,67 @@ protected static function calculate_refund_date( $order, $previous_refund = null
...
+		if ( isset( $args['refund-ratio'] ) && isset( $args['status'] ) && 'completed' === $args['status'] ) {
+			$refund_ratio = floatval( $args['refund-ratio'] );
+			$refund_ratio = max( 0.0, min( 1.0, $refund_ratio ) );
+
+			$total_refunds = (int) round( $count * $refund_ratio );
+
+			// Split refunds: 50% full, 25% single partial, 25% multi-partial
+			$num_full = (int) round( $total_refunds * 0.5 );
+			$num_partial = (int) round( $total_refunds * 0.25 );
+			$num_multi = $total_refunds - $num_full - $num_partial; // Remainder goes to multi
```
[the remark this answers is not in the document]
Comment (layoutd) [reply includes/Generator/Order.php]: Changed in 9d6247717f47a660baacf2daccdeec99914df583

File: includes/Generator/Order.php [hunk 2]
```diff
@@ -121,20 +145,26 @@ public static function generate( $save = true, $assoc_args = array() ) {
 
 		// Handle --coupon-ratio parameter
 		if ( isset( $assoc_args['coupon-ratio'] ) ) {
-			$coupon_ratio = floatval( $assoc_args['coupon-ratio'] );
+			// Use exact ratio flag if in batch mode
+			if ( null !== self::$batch_coupon_flags && ! empty( self::$batch_coupon_flags ) ) {
+				$include_coupon = array_shift( self::$batch_coupon_flags );
```
[the remark this answers is not in the document]
Comment (layoutd) [reply includes/Generator/Order.php]: Changed in 201d302094a69a7c05623d7690a085bdf3932fc2

File: includes/Generator/Order.php [hunk 3]
```diff
@@ -180,30 +210,44 @@ public static function generate( $save = true, $assoc_args = array() ) {
...
-				$refund_ratio = floatval( $assoc_args['refund-ratio'] );
+				$refund_type = self::REFUND_TYPE_NONE;
 
-				// Validate ratio is between 0.0 and 1.0
-				if ( $refund_ratio < 0.0 || $refund_ratio > 1.0 ) {
-					$refund_ratio = max( 0.0, min( 1.0, $refund_ratio ) );
-				}
+				// Use exact ratio flag if in batch mode
+				if ( null !== self::$batch_refund_flags && ! empty( self::$batch_refund_flags ) ) {
+					$refund_type = array_shift( self::$batch_refund_flags );
```
[the remark this answers is not in the document]
Comment (layoutd) [reply includes/Generator/Order.php]: Changed in 201d302094a69a7c05623d7690a085bdf3932fc2

File: includes/Generator/Order.php [hunk 4]
```diff
@@ -790,4 +874,73 @@ protected static function calculate_refund_date( $order, $previous_refund = null
...
+		// Initialize refund flags if refund-ratio is set
+		if ( isset( $args['refund-ratio'] ) && isset( $args['status'] ) && 'completed' === $args['status'] ) {
+			$refund_ratio = floatval( $args['refund-ratio'] );
+			$refund_ratio = max( 0.0, min( 1.0, $refund_ratio ) );
+
+			$total_refunds = (int) round( $count * $refund_ratio );
+
+			// Split refunds: 50% full, 25% single partial, 25% multi-partial
+			$num_full = (int) round( $total_refunds * self::REFUND_DISTRIBUTION_FULL_RATIO );
+			$num_partial = (int) round( $total_refunds * self::REFUND_DISTRIBUTION_PARTIAL_RATIO );
```
[the remark this answers is not in the document]
Comment (layoutd) [reply includes/Generator/Order.php]: Changed in a033f04eb9b7638fdb92371c30ed8c2fdac2722f

File: includes/Generator/Order.php [hunk 5]
```diff
@@ -790,4 +874,73 @@ protected static function calculate_refund_date( $order, $previous_refund = null
...
+				array_fill( 0, $num_with_coupons, true ),
+				array_fill( 0, $num_without, false )
+			);
+
+			// Shuffle for randomness
+			shuffle( self::$batch_coupon_flags );
+		}
+
+		// Initialize refund flags if refund-ratio is set
+		if ( isset( $args['refund-ratio'] ) && isset( $args['status'] ) && 'completed' === $args['status'] ) {
```
[the remark this answers is not in the document]
Comment (layoutd) [reply includes/Generator/Order.php]: Applied in da23c522d7054d2c4e6adeee68c60cf14241f4b9

File: includes/Generator/Order.php [hunk 6]
```diff
@@ -790,4 +874,73 @@ protected static function calculate_refund_date( $order, $previous_refund = null
...
+		self::$batch_coupon_index = 0;
+		self::$batch_refund_index = 0;
+
+		// Initialize coupon flags if coupon-ratio is set
+		if ( isset( $args['coupon-ratio'] ) ) {
+			$coupon_ratio = floatval( $args['coupon-ratio'] );
+			$coupon_ratio = max( 0.0, min( 1.0, $coupon_ratio ) );
+
+			$num_with_coupons = (int) round( $count * $coupon_ratio );
+			$num_without = $count - $num_with_coupons;
```
[the remark this answers is not in the document]
Comment (layoutd) [reply includes/Generator/Order.php]: Seems like that size of request is fairly unlikely. But documented and set to fall back to random assignment at > 10k orders being generated in eeed7d52def32ed9a3a1baeecbd5f12bf67b2763

File: includes/Generator/Order.php [hunk 7]
```diff
@@ -180,30 +235,45 @@ public static function generate( $save = true, $assoc_args = array() ) {
...
 			if ( isset( $assoc_args['refund-ratio'] ) && 'completed' === $status ) {
-				$refund_ratio = floatval( $assoc_args['refund-ratio'] );
+				$refund_type = self::REFUND_TYPE_NONE;
 
-				// Validate ratio is between 0.0 and 1.0
-				if ( $refund_ratio < 0.0 || $refund_ratio > 1.0 ) {
-					$refund_ratio = max( 0.0, min( 1.0, $refund_ratio ) );
-				}
+				// Use exact ratio flag if in batch mode
+				if ( null !== self::$batch_refund_flags && isset( self::$batch_refund_flags[ self::$batch_refund_index ] ) ) {
```
[the remark this answers is not in the document]
Comment (layoutd) [reply includes/Generator/Order.php]: Addressed in 65ba0a0

File: includes/Generator/Order.php [hunk 8]
```diff
@@ -121,20 +169,27 @@ public static function generate( $save = true, $assoc_args = array() ) {
 
 		// Handle --coupon-ratio parameter
 		if ( isset( $assoc_args['coupon-ratio'] ) ) {
-			$coupon_ratio = floatval( $assoc_args['coupon-ratio'] );
+			// Use exact ratio flag if in batch mode
+			if ( null !== self::$batch_coupon_flags && isset( self::$batch_coupon_flags[ self::$batch_coupon_index ] ) ) {
+				$include_coupon = self::$batch_coupon_flags[ self::$batch_coupon_index ];
+				self::$batch_coupon_index++;
```
[the remark this answers is not in the document]
Comment (layoutd) [reply includes/Generator/Order.php]: Addressed in https://github.com/woocommerce/wc-smooth-generator/commit/65ba0a0b8512bb11acf73289b2369b0210d709ac

**Review by chihsuan — APPROVED** [review]
Great work on this feature! @layoutd  Confirmed that coupons/refunds are generated correctly. 👍 

I wanted to add a potential enhancement to this feature. Instead of pre-generating arrays, maybe we can calculate probability dynamically using "selection without replacement", something like this:

```php
  // Initialize counters (O(1) memory)
  $coupons_remaining = isset( $args['coupon-ratio'] )
      ? (int) round( $amount * floatval( $args['coupon-ratio'] ) )
      : 0;
  $orders_remaining = $amount;

  for ( $i = 1; $i <= $amount; $i++ ) {
      $include_coupon = null;

      if ( isset( $args['coupon-ratio'] ) ) {
          // Probability = remaining_coupons / remaining_orders
          // Guarantees exact count while maintaining random distribution
          $include_coupon = ( wp_rand( 1, $orders_remaining ) <= $coupons_remaining );
          if ( $include_coupon ) {
              $coupons_remaining--;
          }
      }

      $orders_remaining--;

      // ... generate order
  }
```

The approach might be worth considering for the elimination threshold.

**Comment by layoutd** [comment]
Thanks for the suggestion @chihsuan, I've implemented the changes to improve . As a follow-up PR, it would be interesting to add unit tests to cover all generation and edge cases.
